### ISSUE TITLE
Known Issue: unknown behaviour for proto.Equal

### ISSUE THREAD

**Issue by awalterschulze** [gogo/protobuf#13]
_From [awalterschulze](https://code.google.com/u/awalterschulze/) on May 07, 2014 21:17:58_

proto.Equal(a, b Message) bool has not been tested with all gogoprotobuf scenarios.

If you are using this please comment so I can up the priority.

_Original issue: http://code.google.com/p/gogoprotobuf/issues/detail?id=13_

**Comment by dennwc** [comment]
Found an issue with this combination of options:
```proto
message M {
    repeated bytes arr = 1 [(gogoproto.customtype) = "MyType", (gogoproto.nullable) = false];
}
```
It will generate next code for M:
```Go
type M struct {
    Arr []MyType `protobuf:"bytes,1,rep,name=arr,customtype=MyType" json:"arr"`
}

func (this *M) Equal(that interface{}) bool {
    // ...
    that1, ok := that.(*M)
    if !ok {
        return false
    }
    // ...
    if !this.Arr[i].Equal(that1.Arr[i]) { // should be &that1.Arr[i]
        return false
    }
    // ...
}
```
It will fail because `Equal` method will try to cast a parameter to `*MyType`.

**Comment by awalterschulze** [comment]
This should be reported in another issue, but you could check out in test/thetest.proto
```
message NidRepCustom {
	repeated bytes Id = 1 [(gogoproto.customtype) = "Uuid", (gogoproto.nullable) = false];
	repeated bytes Value = 2 [(gogoproto.customtype) = "github.com/gogo/protobuf/test/custom.Uint128", (gogoproto.nullable) = false];
}
```
The equal method does the following
```
func (uuid Uuid) Equal(other Uuid) bool {
	return bytes.Equal(uuid[0:], other[0:])
}
```
Does this help?

**Comment by dennwc** [comment]
Yes, it will definitely would, but this will not work for auto-generated "custom" type. I want to point out, that behavior of this function is not symmetric: it will pass `Arr[i]`, but auto-generated variant of `Equal` always assumes a pointer to the type. So maybe it should pass pointers always, or as an alternative, auto-generated `Equal` functions should check for a non-pointer argument also.

**Comment by awalterschulze** [comment]
Could you please create a new issue for this?  Before we start having a longer conversation.

**Comment by andreimatei** [comment]
I've run into problems with `proto.Equal()` and a proto where we use `gogoproto.customtype` for a `bytes` field. I hope this is the right place to report.

This is the proto:
https://github.com/cockroachdb/cockroach/blob/befea82/pkg/roachpb/metadata.proto#L73

And the error is:
```
panic: interface conversion: interface {} is roachpb.RKey, not []uint8
```

It's because of this special handling of `bytes` fields in `proto.Equal`:
https://github.com/golang/protobuf/blob/ef00c02/proto/equal.go#L218


Any ideas for a work-around?

**Comment by awalterschulze** [comment]
You are welcome to contribute a fix with a test, but an easier fix might be to just generate equal methods.
Generating equal methods can be done in two ways:
  1) Using the gogoprotobuf equal extension https://godoc.org/github.com/gogo/protobuf/plugin/equal
  2) Using my new project https://github.com/awalterschulze/goderive

**Comment by andreimatei** [comment]
Thanks! The equal extension looks good! But I've run into another issue with it, reported in #283.

**Comment by tigrannajaryan** [comment]
When using a customtype that is an array proto.Equal does not work, outputs log message `proto: don't know how to compare ...`.

The `switch` at https://github.com/gogo/protobuf/blob/5628607bb4c51c3157aacc3a50f0ab707582b805/proto/equal.go#L151 does not appear to handle `reflect.Array` case which explains why this does not work.

Before digging into the code I also tried defining `Equal` func for my customtype since I was expecting that it may be used, but `proto.Equal` does not appear to try to use it at all.
